applications/luci-app-xfrpc/htdocs/luci-static/resources/view/xfrpc.js

```diff
@@ -0,0 +1,229 @@
+'use strict';
+'require view';
+'require ui';
+'require form';
+'require rpc';
+'require tools.widgets as widgets';
+
+//	[Widget, Option, Title, Description, {Param: 'Value'}],
+var startupConf = [
+	[form.Flag, 'disabled', _('Disabled xfrpc service')],
+	[form.ListValue, 'loglevel', _('Log level'), _('LogLevel specifies the minimum log level. Valid values are "debug", "info", "notice", "warning", "err", "crit", "alert" and "emerg".<br />By default, this value is "info".'), {values: ['7', '6', '5',  '4', '3', '2', '1', '0']}],
+];
+
+var commonConf = [
+	[form.Value, 'server_addr', _('Server address'), _('ServerAddr specifies the address of the server to connect to.<br />By default, this value is "0.0.0.0".'), {datatype: 'host'}],
+	[form.Value, 'server_port', _('Server port'), _('ServerPort specifies the port to connect to the server on.<br />By default, this value is 7000.'), {datatype: 'port'}],
+	[form.Value, 'token', _('Token'), _('Token specifies the authorization token used to create keys to be sent to the server. The server must have a matching token for authorization to succeed. <br />By default, this value is "".')],
+];
+
+var baseProxyConf = [
+	[form.ListValue, 'type', _('Proxy type'), _('ProxyType specifies the type of this proxy. Valid values include "tcp", "http", "https".<br />By default, this value is "tcp".'), {values: ['tcp', 'http', 'https']}],
+	[form.Value, 'local_ip', _('Local IP'), _('LocalIp specifies the IP address or host name to proxy to.'), {datatype: 'ipaddr'}],
+	[form.Value, 'local_port', _('Local port'), _('LocalPort specifies the port to proxy to.'), {datatype: 'port'}],
+];
+
+var bindInfoConf = [
+	[form.Value, 'remote_port', _('Remote port'), _('If remote_port is 0, frps will assign a random port for you'), {datatype: 'port'}]
+];
+
+var domainConf = [
+	[form.Value, 'custom_domains', _('Custom domains')],
+];
```

Comment:
Do we need this abstraction?
Can't we just write down every configuration option.
Then the source of this app looks just like any other app in this repository.

Then we can delete these functions:

- setParams
- defTabOpts
- defOpts

and use the normal API functions from the LuCI.
This makes it easier to read and change the code so that it feels the same to other developers who have already done something in the LuCI.